### PR TITLE
Bugfix: retrieve nested link

### LINKED ISSUE #203: Dependency wheel does not search `Link` attribute definitions nested inside an `Object`
**Describe the bug**

The dependency wheel lists components that can be created based on the selected component's `Link` attribute definitions.

However, if the `Link` attribute definition is nested inside an `Object`, the dependency wheel won't consider it.

**To Reproduce**

https://github.com/ditrit/leto-modelizer/assets/20184007/2d088de8-caa0-43b6-849a-750b26ec0e47

Use this sample definition for the terrator-plugin:

```json
{
  "type": "aws_security_group_nested_link",
  "model": "DefaultModel",
  "icon": "AwsSecurityGroup",
  "attributes": [
    {
      "name": "my_object",
      "type": "Object",
      "attributes": [
        {
          "name": "vpc_id",
          "type": "Link",
          "linkType": "Default",
          "linkRef": "aws_vpc"
        }
      ]
    }
  ]
}
```

There are no errors showing in the console.

**Expected behavior**

The dependency wheel should find `Link` attribute definitions, even if they are nested inside an `Object`.

**Application**

- leto version `1.2.0`

### LINKED ISSUE #208: Remove nested link
**Describe the bug**

When you have a nested link.

You can't remove link by clicking on it or the link is not remove when you remove source or target of the link.

### PR DESCRIPTION
**Pull request by Zorin95670**
Fix #203 
Fix #208

### CODE REVIEW

File: src/models/Component.js [hunk 1]
```diff
@@ -103,6 +103,37 @@ class Component extends FileInformation {
     }
   }
 
+  /**
+   * Create all parents attributes from definitions.
+   * @param {ComponentAttributeDefinition[]} definitions - All parents attributes order from most
+   * depth to lower depth.
```
Comment (justinehell) [inline src/models/Component.js:109]: from most depth to lower depth -> from greatest to least depth.

File: src/models/Component.js [hunk 2]
```diff
@@ -103,6 +103,37 @@ class Component extends FileInformation {
...
+   * @returns {ComponentAttribute[]} - Deeper created parent.
```
Comment (justinehell) [inline src/models/Component.js:110]: Deeply created parent.

File: src/models/Component.js [hunk 3]
```diff
@@ -103,6 +103,37 @@ class Component extends FileInformation {
...
+   * @private
+   */
+  __createCascadeAttributes(definitions) {
```
Comment (justinehell) [inline src/models/Component.js:113]: cascade -> nested ?

File: src/models/Component.js [hunk 4]
```diff
@@ -157,17 +225,27 @@ class Component extends FileInformation {
...
-        if (attribute.definition && attribute.definition.type === 'Link') {
-          return attribute.removeLink(id);
-        }
+    this.__removeLinkAttribute(this.attributes, id, name);
+  }
 
-        return true;
-      });
+  /**
+   * Remove all attributes refer to specify link.
```
Comment (justinehell) [inline src/models/Component.js:232]: Remove all attributes related to the specified link ?

File: src/models/Component.js [hunk 5]
```diff
@@ -157,17 +225,27 @@ class Component extends FileInformation {
...
+   * @param {ComponentAttribute[]} attributes - Attributes to find attribute link and delete link
+   * reference.
+   * @param {string} id - Id of the target link.
+   * @param {string} name - Attribute name that store link ids.
```
Comment (justinehell) [inline src/models/Component.js:236]: store -> stores

File: src/models/Component.js [hunk 6]
```diff
@@ -204,6 +282,43 @@ class Component extends FileInformation {
...
+      }
+
+      if (this.__compareAttributeField(attribute, field, ...values)) {
+        result.push(attribute);
+      }
+    });
+  }
+
+  /**
+   * Compare attribute field to given value and indicate if its contains or not.
```
Comment (justinehell) [inline src/models/Component.js:307]: ..indicate if the value contains that attribute or not.

File: src/models/Component.js [hunk 7]
```diff
@@ -204,6 +282,43 @@ class Component extends FileInformation {
...
+   * @param {ComponentAttribute} attribute - Attribute to check.
+   * @param {string} field - Name of attribute field.
+   * @param {...string} values - Valid field value to search in attribute field.
+   * @returns {boolean} True if field value is containing in provided values.
```
Comment (justinehell) [inline src/models/Component.js:311]: True if the value of the field is contained in the provided values.

File: tests/unit/models/Component.spec.js [hunk 8]
```diff
@@ -333,6 +334,128 @@ describe('Test class: Component', () => {
...
+          }),
+        });
+
+        const currentAttributes = component.__createCascadeAttributes([]);
+
+        expect(component.attributes).toEqual([]);
+        expect(currentAttributes).toEqual([]);
+      });
+
+      it('Should create root attribute if not exists', () => {
```
Comment (justinehell) [inline tests/unit/models/Component.spec.js:351]: ...if it doesn't exist

File: tests/unit/models/Component.spec.js [hunk 9]
```diff
@@ -333,6 +334,128 @@ describe('Test class: Component', () => {
...
+          definition: rootAttributeDefinition,
+        });
+
+        const currentAttributes = component.__createCascadeAttributes([rootAttributeDefinition]);
+
+        expect(component.attributes).toEqual([expectedAttribute]);
+        expect(currentAttributes).toEqual([]);
+      });
+
+      it('Should create root and sub attribute if no one exists', () => {
```
Comment (justinehell) [inline tests/unit/models/Component.spec.js:375]: ...if there is none

File: package-lock.json [hunk 10]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "leto-modelizer-plugin-core",
   "version": "0.18.0",
-  "lockfileVersion": 3,
+  "lockfileVersion": 2,
```
Comment (justinehell) [inline package-lock.json:4]: no `package.json` file updated ?

File: changelog.md [hunk 11]
```diff
@@ -13,13 +13,16 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
-
 - Add `arrangeComponentsPosition` method in `DefaultDrawer` and `DefaultPlugin`.
 - Add inheritable class `DefaultLayout` for automatic layout in the diagram (does nothing).
 - Add inherited class `ElkLayout` for automatic layout in the diagram, using ELK.
 - Add HTML attribute equality as a Cypress step definition in `html.js`.
 - Add `elkjs` and `web-worker` as package dependencies.
 
+### Fixed
+
+- `getDefinedAttributesByType` will now return nested attribute inside object, fix [this bug](https://github.com/ditrit/leto-modelizer-plugin-core/issues/203).
```
Comment (justinehell) [inline changelog.md:24]: attributes ?

**Review by justinehell — CHANGES_REQUESTED** [review]

File: src/models/Component.js (hunk 1 above)
Comment (Zorin95670) [reply src/models/Component.js:109]: done

File: src/models/Component.js (hunk 2 above)
Comment (Zorin95670) [reply src/models/Component.js:110]: done

File: src/models/Component.js (hunk 3 above)
Comment (Zorin95670) [reply src/models/Component.js:113]: done

File: src/models/Component.js (hunk 4 above)
Comment (Zorin95670) [reply src/models/Component.js:232]: done

File: src/models/Component.js (hunk 5 above)
Comment (Zorin95670) [reply src/models/Component.js:236]: done

File: src/models/Component.js (hunk 6 above)
Comment (Zorin95670) [reply src/models/Component.js:307]: done

File: src/models/Component.js (hunk 7 above)
Comment (Zorin95670) [reply src/models/Component.js:311]: done

File: tests/unit/models/Component.spec.js (hunk 8 above)
Comment (Zorin95670) [reply tests/unit/models/Component.spec.js:351]: done

File: tests/unit/models/Component.spec.js (hunk 9 above)
Comment (Zorin95670) [reply tests/unit/models/Component.spec.js:375]: done

File: changelog.md (hunk 11 above)
Comment (Zorin95670) [reply changelog.md:24]: done

**Review by aboubakar14 — APPROVED** [review]

**Pull request merged**